### PR TITLE
Provides fs statistics

### PR DESCRIPTION
**Pull request by mariuszbarczak**

### CODE REVIEW

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+}
+
+void FilesystemStatistics::getFilesystemStatistics(
+        proto::FilesystemStatistics *statistics) const {
+    fill(statistics, "");
+}
+
+FilesystemStatistics &FilesystemStatistics::get(
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.cpp]: Please also add short comment about usage. For example:

"FilesystemStatistics are organized in a tree. To get statistics of an element knowing only it's id, we need to recursively get the parentId of id until we get the id of root element. Then we can find the requested element by going down the tree by converting the ids to names of the tree elements."

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        } else {
+            entry.set_directory(dir);
+        }
+
+        if (checkWif) {
+            auto &metrics = *entry.mutable_statistics()
+                                     ->mutable_write()
+                                     ->mutable_metrics();
+
+            if (metrics[WIF_METRIC_NAME].value() <= 1.0) {
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.cpp]: Add a comment about this logic please

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    }
+}
+
+void FilesystemStatistics::getFilesystemStatistics(
+        proto::FilesystemStatistics *statistics) const {
+    fill(statistics, "");
+}
+
+FilesystemStatistics &FilesystemStatistics::get(
+        IFileSystemViewer *viewer,
```
[the remark this answers is not in the document]
Comment (mariuszbarczak) [reply source/octf/analytics/statistics/FilesystemStatistics.cpp]: We can not, because reference has to be set when defining.

**Comment by tomaszrybicki** [comment]
retest this please

File: source/octf/analytics/statistics/FilesystemStatistics.h [hunk 4]
```diff
@@ -45,30 +45,38 @@ class FilesystemStatistics {
...
+                                             uint64_t partId);
 
-    void count(const proto::trace::ParsedEvent &event);
+    void fillProtoStatistics(proto::FilesystemStatistics *statistics,
+                             const std::string &dir) const;
+
+    void fillProtoStatisticsEntry(
+            proto::FilesystemStatisticsEntry *entry) const;
+
+    void updateIoStats(const proto::trace::ParsedEvent &event);
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.h]: 👍

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 5]
```diff
@@ -111,16 +178,17 @@ void FilesystemStatistics::fill(proto::FilesystemStatistics *statistics,
                                      ->mutable_write()
                                      ->mutable_metrics();
 
-            if (metrics[WIF_METRIC_NAME].value() <= 1.0) {
-                continue;
+            if (metrics[WIF_METRIC_NAME].value() > 1.0l) {
+                // To minimize output, copy statistics with a invalidation only
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.cpp]: 1) 'an invalidation'
2) This comment seems unfinished

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 6]
```diff
@@ -156,23 +226,41 @@ void FilesystemStatistics::fill(proto::FilesystemStatisticsEntry *entry,
     double written =
             pStats->write().size().total() + pStats->discard().size().total();
 
-    double wif = written / workset;
+    double wif = 0.0;
+    if (workset != 0.0l) {
+        wif = written / workset;
+    }
+
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.cpp]: What's the logic behind this check? Shouldn't it be  '!= 0.0' ?

File: source/octf/analytics/statistics/FilesystemStatistics.cpp [hunk 7]
```diff
@@ -111,16 +178,17 @@ void FilesystemStatistics::fill(proto::FilesystemStatistics *statistics,
                                      ->mutable_write()
                                      ->mutable_metrics();
 
-            if (metrics[WIF_METRIC_NAME].value() <= 1.0) {
-                continue;
+            if (metrics[WIF_METRIC_NAME].value() > 1.0l) {
```
Comment (tomaszrybicki) [inline source/octf/analytics/statistics/FilesystemStatistics.cpp]: Why not "value > 1" ?

**Pull request merged**